Item Bugs

t-moe opened the pull request:
This MR fixes 3 bugs (#6, #7 and #17), the most critical being #6. I also fixed some Memory Leaks which I found during bugfixing.

When reviewing this, you should also review if the new cleanup strategy makes sense.

> New behaviour:
> * If item's are deleted by the user (via contextmenu or pressing delete) the inputs/output connections will be removed and the corresponding signal/slots will be fired.
> * In the item's destructor (which is only invoked directly if the project get's closed or the item has no scene) the inputs/outputs will be deleted but no signals will be fired.

Commit:
Fixed cleanup of item input&outputs. Closes #6.

New behaviour:
* If item's are deleted by the user (via contextmenu or pressing delete) the inputs/output connections will be removed and the corresponding signal/slots will be fired.
* In the item's destructor (which is only invoked directly if the project get's closed or the item has no scene) the inputs/outputs will be deleted but no signals will be fired.

Commit:
Fixed some memory leaks in GuiManager, PluginManager and ItemToolbox

Commit:
Added raise() to AbstractWindowItem::openWindow. Closes #7

Commit:
Renaming Window Title on Item Name change. Closes #17

fschlaepfer commented on src/item/abstract_item.cpp:
```diff
@@ -212,7 +212,7 @@ void AbstractItem::contextMenuEvent(QGraphicsSceneContextMenuEvent* event)
     if (actionSel == actionRemove) {
-        remove();
+        disconnectConnections();
         delete this;
```
We should set `this = nullptr` after this to prevent accessing it after the memory has been freed.

t-moe replied:
[`this` cannot be assigned.](http://stackoverflow.com/questions/13476879/how-can-you-assign-a-value-to-the-pointer-this-in-c)

fschlaepfer commented on src/item/abstract_item.cpp:
```diff
@@ -450,14 +450,28 @@ QPen AbstractItem::connectorStyle(int type)
-void AbstractItem::remove()
+void AbstractItem::disconnectConnections() {
+    Q_D(AbstractItem);
+    for (int i = d->_inputs.count() - 1; i >= 0; i--) {
```
I'd suggest using `std::for_each` with `QList::rbegin` to prevent errors related to the index variable `i` and also to make the code more easily understandable.

Maybe also add a comment explaining why this needs to be done in reverse order.

fschlaepfer commented on src/item/abstract_window_item.cpp:
```diff
@@ -7,6 +7,10 @@
 AbstractWindowItem::AbstractWindowItem(QString const& typeName) :
     AbstractItem(typeName), d_ptr(new AbstractWindowItemPrivate(this))
 {
+    connect(this,&AbstractItem::nameChanged,this,[this](){
```
We could also move the `Q_D(...)` before the lambda definition and only capture `d` instead of `this` to signal that no public member functions are used in the lambda. As soon as you add any code after the lambda you'll probably also need access to `d` outside the lambda anyway.

fschlaepfer commented on src/item/item_connector.cpp:
```diff
@@ -24,32 +24,33 @@ Item_Connector::Item_Connector(ItemOutput* output, ItemInput* input)
 
-void Item_Connector::remove_and_delete()
+
+void Item_Connector::checkConnection()
```
Since this may delete the `this` object, let's rename it to indicate as much. I feel like a slot named `checkConnection` should not delete the object.

fschlaepfer commented on src/item/item_connector.cpp:
```diff
@@ -24,32 +24,33 @@ Item_Connector::Item_Connector(ItemOutput* output, ItemInput* input)
 {
-    if (input->output() == output) {
-        input->disconnectOutput();
+    if (input->output() != output || !output->inputs().contains(input)) {
```
I find it quite difficult to make sense of this condition. Rewriting it as `!(input->output() == output && output->inputs().contains(input))` makes it a bit easier to understand. Perhaps even extract the condition into a lambda predicate to give it a useful name.

fschlaepfer commented on src/item/item_note.cpp:
```diff
@@ -234,7 +234,7 @@ void ItemNote::contextMenuEvent(QGraphicsSceneContextMenuEvent* event)
 
     if (sel_action == act_remove) {
-        remove();
+        delete this;
```
Same here, let's set `this = nullptr` afterwards to prevent nasty runtime errors.

fschlaepfer commented on src/item/item_toolbox.cpp:
```diff
@@ -14,7 +14,7 @@ Item_Toolbox::Item_Toolbox(QWidget* parent) :
 {
     ui->setupUi(this);
-    model = new Item_List_Model(parent);
+    model = new Item_List_Model(parent==nullptr?this:parent);
```
Have you considered using `parent ? parent : this` instead?

Commit:
Some readability changes in the Item Class code.
Related to #20

t-moe commented:
Thanks for reviewing @fschlaepfer . I fixed most of your addressed concerns (we can discuss the others in the office).
I've merged this now, because @Shadowigor needs this fix as well and I don't want to wait another week :grin:
